coordinatorlib/src/main/scala/com/socrata/datacoordinator/secondary/PlaybackToSecondary.scala

```diff
@@ -185,6 +183,176 @@ class PlaybackToSecondary[CT, CV](u: PlaybackToSecondary.SuperUniverse[CT, CV],
+      if(!it.hasNext || it.next() != Delogger.RowsChangedPreview) {
+        return false
+      }
+      if(it.hasNext && it.head == Delogger.Truncated) {
+        it.next()
+      }
+      while(it.hasNext && it.head == Delogger.RowDataUpdated) {
+        it.next()
+      }
+      if(!it.hasNext || it.next() != Delogger.LastModifiedChanged) {
+        return false
+      }
+
+      !it.hasNext
+    }
+
+    def consolidate(delogger: Delogger[CV], startingDataVersion: Long, endingDataVersion: Long) =
+      new Iterator[Delogger.LogEvent[CV]] with AutoCloseable {
+        // ok, so we're going to want to produce a fake event stream
+        // shaped like a consolidatable (see above) version.  To do
+        // this, we'll combine all the versions' RowsChangedPreview
+        // events then the concatenation of their RowDataUpdateds,
+        // then the final job's LastModifiedChanged.
+
+        val rs = new ResourceScope
+
+        val (effectiveStartingDataVersion, fakeRCP) =
+          (startingDataVersion to endingDataVersion).foldLeft((startingDataVersion, Delogger.RowsChangedPreview(0, 0, 0, false))) { (acc, i) =>
+            val (currentStartingDataVersion, rcp) = acc
+            managed(delogger.delog(i)).run(_.buffered.headOption) match {
+              case Some(Delogger.RowsChangedPreview(rowsInserted, rowsUpdated, rowsDeleted, false)) =>
+                (currentStartingDataVersion, Delogger.RowsChangedPreview(rcp.rowsInserted + rowsInserted,
+                                                                         rcp.rowsUpdated + rowsUpdated,
+                                                                         rcp.rowsDeleted + rowsDeleted,
+                                                                         rcp.truncated))
+              case Some(rcp : Delogger.RowsChangedPreview) =>
+                // it was truncated; just start from here
+                (i, rcp)
+              case _ =>
+                throw ResyncSecondaryException("Consolidation saw the log change?!  No RowsChangedPreview!")
+            }
+          }
+
+        // bleargghhgh -- annoying that this super lazy flatmap over
+        // the data versions has a side effect, but I can't think of a
+        // clearer way to do this.
+        var mostRecentLastModified: Option[Delogger.LastModifiedChanged] = None
+        val data = (effectiveStartingDataVersion to endingDataVersion).iterator.flatMap { i =>
+          new Iterator[Delogger.LogEvent[CV]] {
+            var done: Boolean = false
+
+            val eventsRaw = rs.open(delogger.delog(i))
+            val events = eventsRaw.buffered
+            if(!events.hasNext || events.head.companion != Delogger.RowsChangedPreview) {
+              throw ResyncSecondaryException("Consolidation saw the log change?!  First item was gone or not RowsChangedPreview!")
+            }
+            if(events.next().asInstanceOf[Delogger.RowsChangedPreview].truncated) {
```

Comment:
i don't think i understand this cast..

Reply:
`truncated` is a field on `Delogger.RowsChangedPreview`, which is a subclass of `Delogger.Event` - in the previous conditional we've checked that it both exists and _is_ a RCP, so the downcast is safe.

Reply:
(this would be clearer to write with `nextOption` and a `match`, but `nextOption` doesn't actually exist on `Iterator`)